crates/red_knot/src/semantic/types.rs

```diff
@@ -236,18 +247,34 @@ impl TypeStore {
             .add_class(name, scope_id, bases)
     }
 
-    fn add_union(&self, file_id: FileId, elems: &[Type]) -> UnionTypeId {
+    fn add_class(&self, file_id: FileId, name: &str, scope_id: ScopeId, bases: Vec<Type>) -> Type {
+        Type::Class(self.add_class_type(file_id, name, scope_id, bases))
+    }
+
+    /// add "raw" union type with exactly given elements
+    fn add_union_type(&self, file_id: FileId, elems: &[Type]) -> UnionTypeId {
+        self.add_or_get_module(file_id).add_union(elems)
+    }
+
+    /// add union with normalization; may not return a `UnionType`
+    fn add_union(&self, file_id: FileId, elems: &[Type]) -> Type {
         let mut flattened = Vec::with_capacity(elems.len());
         for ty in elems {
             match ty {
                 Type::Union(union_id) => flattened.extend(union_id.elements(self)),
                 _ => flattened.push(*ty),
             }
         }
-        self.add_or_get_module(file_id).add_union(&flattened)
+        // TODO de-duplicate union elements
```

Comment:
We probably also want to avoid creating `UnionIds` for two unions that are identical. But let's solve this another day

Reply:
Yes, that's part of union normalization in my mind, but I guess there's no TODO for it here; I'll add one.